File: Axiom/Internal/Frames/Temp.cs
namespace Axiom.Internal.Frames
{
    internal class Temp
    {
        private static int count;

        public Temp(string name)
        {
            Name = name;
        }

        public override bool Equals(object obj)
        {
            return Name == ((Temp)obj).Name;
        }

        public static bool operator ==(Temp lhs, Temp rhs)
        {
            return lhs.Equals(rhs);
        }

        public static bool operator !=(Temp lhs, Temp rhs)
        {
            return !lhs.Equals(rhs);
        }

        public override int GetHashCode()
        {
            return Name.GetHashCode();
        }

        public string Name { get; private set; }

        public static Temp Create()
        {
            return new Temp("T" + (count++).ToString());
        }

        public static Temp Create(string name)
        {
            return new Temp("_" + name);
        }
    }
}

File: src/main/java/com/moybl/axiom/frame/Temp.java
package com.moybl.axiom.frame;

public class Temp {

	private String name;

	public Temp(String name) {
		this.name = name;
	}

	public String getName() {
		return name;
	}

	@Override
	public boolean equals(Object obj) {
		if (obj == null || obj.getClass() != this.getClass()) {
			return false;
		}

		return name.equals(((Temp) obj).name);
	}

	private static int counter;

	public static Temp create() {
		counter++;
		return new Temp("T" + counter);
	}

}
